### PR TITLE
Adaptation to use embbeded JRE in bundled applications made with Max

### PR DESCRIPTION
**Pull request by kristofb**
﻿I adapted the a little the code from Peter, in order to be able to merge on the latest commit in master.
I adapted the code so now, JRE 8+ are accepted, by setting proper libjli.dylib path.
This enables:

    Max 8 beta runtime to find and use JRE 8+ stored in max-mxj package directory
    (Max 8 doesn't search for java outside of mxj, so its fine to detect JRE)

This doesn't enable:

    Use of old JRE 6 as embedded JRE. But it doesn't change the behaviour of mxj when nothing is embedded.

Comments from Peter Slack (I used his merge request as a base to start mine), still valid:
I've added an option to embedd a jre directly in the max-mxj packages file for OSX
the directory jre should be placed in max-mxj package directory
it should point to the base of an installed jre or copy the jre into this directory
the option looks for jre/Contents/Home to exist
this only works for 64 bit mode, there is no 32 bit embedded option for OSX
I would also like to merge in the changes from the latest mxj and windows embedd option will follow

It has been tested on OSX and Windows, with Java 8 and various OpenJDK / OpenJRE up to 15.

### CODE REVIEW

**Comment by jeremybernstein** [comment]
This seems to be working on my local install to the extent that I can test it. I'll need to do some additional testing on Windows. If there is no embedded JRE, where does the object search exactly?

**Comment by pslack** [comment]
I'm unsure about removing 32 bit version from Windows if that was a thing Cycling has done, it makes sense to remove 32 bit but not sure if it makes sense to remove it here unless it was removed from production.

File: source/mxj/mxj_win.c [hunk 1]
```diff
@@ -293,18 +301,29 @@ CreateExecutionEnvironment(
 	/* If we got here, jvmpath has been correctly initialized. */
 	// HACK...the above code is crashing when reading the config file. 
 	// currently assuming we will load the following file. revisit. -jkc
-	sprintf(jvmpath, "%s\\bin\\client\\jvm.dll", jrepath);
+	sprintf(jvmpath, "%s\\bin\\client\\"JVM_DLL, jrepath);
```
Comment (pslack) [inline source/mxj/mxj_win.c:304]: missing closed quotes?

File: source/mxj/mxj_win.c [hunk 2]
```diff
@@ -293,18 +301,29 @@ CreateExecutionEnvironment(
...
 	if (stat(jvmpath, &s) != 0) {
 		// rbs: on my install of the JVM for x64 the jvm.dll was in this path
 		// not sure what the correct way is to get this path, but this will do for 
 		// now
-		sprintf(jvmpath, "%s\\bin\\server\\jvm.dll", jrepath);
+		sprintf(jvmpath, "%s\\bin\\server\\"JVM_DLL, jrepath);
```
Comment (pslack) [inline source/mxj/mxj_win.c:309]: missing closed quotes not sure how this compiles?

File: source/mxj/mxj~.vcxproj [hunk 3]
```diff
@@ -19,35 +19,36 @@
     </ProjectConfiguration>
   </ItemGroup>
   <PropertyGroup Label="Globals">
-    <ProjectGuid>{1E9D97C2-7F5A-4299-AB5A-AD0C596D1FB4}</ProjectGuid>
+    <ProjectGuid>{3EEFC698-6D82-4992-AFF1-807E25D42C72}</ProjectGuid>
```
Comment (pslack) [inline source/mxj/mxj~.vcxproj:22]: not sure if this is wise to change the UUID of the project but may be benign, this is a question for Cycling engineering

File: source/mxj/mxj~.vcxproj [hunk 4]
```diff
@@ -74,13 +75,13 @@
     <OutDir Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">../../externals</OutDir>
     <IntDir Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">../../sysbuild/intermediate/$(Configuration)_$(Platform)/$(ProjectName)/\</IntDir>
     <LinkIncremental Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">false</LinkIncremental>
-    <OutDir Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">../../externals</OutDir>
+    <OutDir Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">Z:\Logelloop5-Win-Shared\Logelloop 5 For Windows\resources\packages\max-mxj\externals\</OutDir>
```
Comment (pslack) [inline source/mxj/mxj~.vcxproj:78]: not sure why hard coded path is here and replacing the original relative path

**Comment by jeremybernstein** [comment]
I'm not concerned about 32-bit support. Since 32-bit versions apparently don't support embedding at all, users can grab old versions of the package if they need 32-bit binaries. I'm much more concerned about making real Cmake projects and getting arm64/universal 64-bit compiles working (and solving any arm64 problems) at this point.

File: source/maxsdk/source/maxmspsdk.xcconfig [hunk 5]
```diff
@@ -11,8 +11,8 @@
 
 // Name & Version
 PRODUCT_NAME = $(PROJECT_NAME)
-PRODUCT_VERSION = 7.0.1
-ARCHS = i386 x86_64
```
Comment (pslack) [inline source/maxsdk/source/maxmspsdk.xcconfig:15]: not sure why 32 bit is removed

File: source/mxj/mxj.sln [hunk 6]
```diff
@@ -1,25 +1,40 @@
...
-		HideSolutionNode = FALSE
-	EndGlobalSection
-EndGlobal
+Microsoft Visual Studio Solution File, Format Version 12.00
+# Visual Studio 15
+VisualStudioVersion = 15.0.28307.168
+MinimumVisualStudioVersion = 10.0.40219.1
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "mxj", "mxj.vcxproj", "{1E9D97C2-7F5A-4299-AB5A-AD0C596D1FB4}"
+EndProject
+Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "mxj~", "Z:\Win-8_1-SharedFolder\maximus-dev\max-mxj\source\mxj\mxj~.vcxproj", "{3EEFC698-6D82-4992-AFF1-807E25D42C72}"
```
Comment (pslack) [inline source/mxj/mxj.sln:7]: should be relative path

File: source/mxj/mxj.vcxproj [hunk 7]
```diff
@@ -21,33 +21,34 @@
   <PropertyGroup Label="Globals">
     <ProjectGuid>{1E9D97C2-7F5A-4299-AB5A-AD0C596D1FB4}</ProjectGuid>
     <RootNamespace>mxj</RootNamespace>
+    <WindowsTargetPlatformVersion>10.0.17763.0</WindowsTargetPlatformVersion>
```
Comment (pslack) [inline source/mxj/mxj.vcxproj:24]: I'm not sure if this target is correct, what about windows 8 or 7 ?

File: CMakeLists.txt [hunk 8]
```diff
@@ -14,22 +14,12 @@ if (APPLE)
 else ()
 	set (MSBUILD "C:\\Program Files (x86)\\MSBuild\\12.0\\Bin\\MSBuild.exe")
 	
-	message(STATUS "Building mxj 32")
```
Comment (pslack) [inline CMakeLists.txt:17]: not sure about removal of 32 bit

File: java-classes/max.java.config.txt [hunk 9]
```diff
@@ -52,10 +52,8 @@ max.jvm.option -Xmx256m
 ; your mxj code using JDB or some other debugger which supports ;
 ; the JDI wire protocol ;
 
-; max.jvm.option -Xdebug  ;
-; max.jvm.option -Xnoagent ;
-; max.jvm.option -Xrunjdwp:transport=dt_socket\,address=8074\,server=y\,suspend=n ; 
-; max.jvm.option -XX:-UseSharedSpaces ;
+; max.jvm.option -XX:-UseSharedSpaces
```
Comment (pslack) [inline java-classes/max.java.config.txt:55]: I suspect these need to end with ; to be a comment

File: source/mxj/JavaHomeOsx.c [hunk 10]
```diff
@@ -43,36 +45,127 @@ int isVMLibrary( _TCHAR* vm );
...
+                }
+            }
+        }
+    }
+    return privateEmbeddedHomeDirectory;
+#endif
+}
 
-/* Define the window system arguments for the various Java VMs. */
-static char*  argVM_JAVA[] = { "-XstartOnFirstThread", NULL };
```
Comment (pslack) [inline source/mxj/JavaHomeOsx.c:67]: this is an important threading issue on OSX . .if it wasn't used then by all means remove it or is it replaced elsewhere

**Review by pslack — COMMENTED** [review]
I feel there are top level directions from Cycling engineering that are missing or need to be updated : what is the minimum target (ie win 8 or what OSX version etc), what are the supported platforms (x86 x86_64, arm64)? With this many changes I feel there also needs to be an explicit test matrix so we can cover all the permutations besides what are the minimum production specs from Cycling.  In the current state we covered 32 and 64 bit platforms for both OSX and Windows and only JDK, today those specifications are different but I feel it isn't up to contributors like myself to decide.  There are some specific tests that were done, for example use with clojure to ensure it meets requirements.

**Comment by jeremybernstein** [comment]
> I feel there are top level directions from Cycling engineering that are missing or need to be updated : what is the minimum target (ie win 8 or what OSX version etc), what are the supported platforms (x86 x86_64, arm64)? With this many changes I feel there also needs to be an explicit test matrix so we can cover all the permutations besides what are the minimum production specs from Cycling. In the current state we covered 32 and 64 bit platforms for both OSX and Windows and only JDK, today those specifications are different but I feel it isn't up to contributors like myself to decide. There are some specific tests that were done, for example use with clojure to ensure it meets requirements.

Oh sorry, that was unclear from me. I don't expect that this pull request satisfies any of my engineering requirements, merely that it implements the embedding feature which has been a pain point for users, especially as Java is removed from the OS on the macOS side. 

I'm planning to do the leg work on what is to follow (real Cmake project, universal/arm64 support, etc.) and I'm totally happy with the state of the work to this point.

File: source/mxj/mxj_win.c (hunk 2 above)
Comment (jeremybernstein) [reply source/mxj/mxj_win.c:309]: it's using C string concatenation, that's fine AFAICT

File: source/mxj/mxj_win.c (hunk 1 above)
Comment (jeremybernstein) [reply source/mxj/mxj_win.c:304]: it's using C string concatenation, that's fine AFAICT (the close quote is between 'client\\' and 'JVM_DLL'

File: source/mxj/mxj.vcxproj [hunk 11]
```diff
@@ -74,13 +75,13 @@
     <OutDir Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">../../externals</OutDir>
     <IntDir Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">../../sysbuild/intermediate/$(Configuration)_$(Platform)/$(ProjectName)/\</IntDir>
     <LinkIncremental Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">false</LinkIncremental>
-    <OutDir Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">../../externals</OutDir>
+    <OutDir Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">Z:\Logelloop5-Win-Shared\Logelloop 5 For Windows\resources\packages\max-mxj\externals\</OutDir>
```
[the remark this answers is not in the document]
Comment (jeremybernstein) [reply source/mxj/mxj.vcxproj:78]: clearly a typo

File: source/mxj/mxj~.vcxproj (hunk 3 above)
Comment (jeremybernstein) [reply source/mxj/mxj~.vcxproj:22]: it's probably benign

File: source/mxj/mxj~.vcxproj (hunk 4 above)
Comment (jeremybernstein) [reply source/mxj/mxj~.vcxproj:78]: typo

File: source/mxj/mxj.sln (hunk 6 above)
Comment (jeremybernstein) [reply source/mxj/mxj.sln:7]: agreed, typo

File: source/mxj/mxj.vcxproj (hunk 7 above)
Comment (jeremybernstein) [reply source/mxj/mxj.vcxproj:24]: I don't know if this is equivalent to "minimum deployment target" on macOS (lowest OS on which the binary will load). If so, we should target Win7.

File: CMakeLists.txt (hunk 8 above)
Comment (jeremybernstein) [reply CMakeLists.txt:17]: This entire file is going to be completely rewritten, but in principle 32-bit disappearing is fine.

File: source/mxj/mxj~.vcxproj (hunk 3 above)
Comment (pslack) [reply source/mxj/mxj~.vcxproj:22]: I gather the project files would be irrelevant with a pure cmake build , I am assuming the product GUID could be something set in the CMake files although I've never tried it for windows.  I guess I recommend to keep the value consistent with previous releases for updates etc.

Comment (jeremybernstein) [reply source/mxj/mxj~.vcxproj:22]: I believe that GUID isn't used for anything relevant to publishing/installing/etc. But I don't have a strong feeling about it, and as you suggest, it's about to be irrelevant.

File: java-classes/max.java.config.txt (hunk 9 above)
Comment (jeremybernstein) [reply java-classes/max.java.config.txt:55]: @kristofb this is true (although the actual implementation makes it less critical), will fix.

File: source/mxj/JavaHomeOsx.c (hunk 10 above)
Comment (jeremybernstein) [reply source/mxj/JavaHomeOsx.c:67]: @pslack  can you expand on this? The reading I've done indicates that it's necessary for SWT applications (but maybe not all the time?) on macOS. In general, I'd support removing this from hard-coded options and permit its use in max.java.config.txt where necessary. Maybe we can add it there with some explanatory text, but commented. @kristofb ?

**Comment by jeremybernstein** [comment]
OK, I've kicked the tires and am comfortable with these changes, which appear to be working properly. I'm prepared to merge this to master once we clear up the usage of `-XstartOnFirstThread`. I have a branch based on this work which adds true Cmake support, and then a branch based on _that_ branch which adds arm64 support, so merging this is a first step to that bit of progress.

File: source/mxj/JavaHomeOsx.c (hunk 10 above)
Comment (jeremybernstein) [reply source/mxj/JavaHomeOsx.c:67]: should be removed and removing it.
